### ISSUE TITLE
Scheduled Source with 0 initialDelay is using interval as initialDelay

### ISSUE THREAD

**Issue by francisdb** [akka/akka#18383]
The test app below with initialDelay = 0.nanos prints this:

```
root 2015-09-01T16:28:12.209Z BEFORE
root 2015-09-01T16:28:17.317Z TICK
```
-> applied initialDelay is ~5 seconds

now change to initialDelay = 1.nanos and you get

```
root 2015-09-01T16:28:38.760Z BEFORE
root 2015-09-01T16:28:38.889Z TICK
```
-> initialDelay is correctly applied

```scala
import java.time.Instant

import akka.actor.ActorSystem
import akka.stream.ActorMaterializer
import akka.stream.scaladsl.{Sink, Source}

import scala.concurrent.duration._

object BugInitialDelay extends App{

  import scala.concurrent.ExecutionContext.Implicits.global

  implicit val system = ActorSystem("foo")
  implicit val materializer = ActorMaterializer()

  println(Instant.now() + " BEFORE")
  Source(initialDelay = 0.nanos, interval = 5.seconds, tick = ()).take(1).map{ tick =>
    println(Instant.now() + " TICK")
  }.runWith(Sink.ignore).onComplete { _ =>
    system.shutdown()
  }
}
```

**Comment by patriknw** [comment]
It is probably because there is no demand from downstream yet. As documented "If a consumer has not requested any elements at the point in time when the tick element is produced it will not receive that tick element later. It will receive new tick elements as soon as it has requested more elements."

**Comment by francisdb** [comment]
So with 0 nanoseconds there is no demand and with 1 nanosecond there is? Does not make sense to me.

**Comment by drewhk** [comment]
@francisdb the resolution of the scheduler is way coarser than a nanosecond. The configuration key "akka.scheduler.tick-duration" controls the granularity, and it is by default 10ms. I.e. it is not 1 nanosecond anymore but 10ms which is more than enough to get the demand from the downstream. Anyway, we can special case the 0 initialDelay to mean "start as soon as there is demand".

**Comment by francisdb** [comment]
@drewhk I don't expect anything to have nanosecond granularity :-)
The problem with having to set initial delays is always 'what is enough?'. A bit like trying to add Thread.sleep(?) to solve concurrency issues.
In this case 1 ns - 10 ms with immediate demand means "start directly" where 0 sec means that initialDelay = interval.
On a high load server is here a chance that immediate demand shifts beyond 10ms?

So with this special case implemented would 0 sec initialDelay and 5 sec interval with demand starting at 3 sec give a tick at 3 sec and one at 5 sec or 3 sec and 9 sec?

**Comment by drewhk** [comment]
> On a high load server is here a chance that immediate demand shifts beyond 10ms?

Of course! When I said "more than enough" I meant as in explaining the phenomena (i.e. "1ns counts as 10ms which is more than enough in general to get the initial demand, which explains what you observed"), not as a recommendation (i.e. "this is more then enough for your use case, use it happily in production").

> In this case 1 ns - 10 ms with immediate demand means "start directly" where 0 sec means that initialDelay = interval.

There are two ways you can interpret this scheduling element:
 - initialDelay is counted form the first demand
 - initialDelay is counted from startup of this stage

Currently, it is the second option that is implemented, and yes, in that case initialDelay will give you a *lower bound* on the actual initial delay, since if there is no demand available, it will wait for one at least interval more before emitting a tick. We can change it so that initialDelay is counted from the first demand.

**Comment by francisdb** [comment]
Thanks @drewhk for clearing that up. Guess I for now will go for a 1s initialDelay just to be sure that in any case things start running when the service boots.

**Comment by drewhk** [comment]
There might be a simpler solution if you want to have a "guaranteed" first element:

    Source.single(initialtick).concat(Source(..ticks...)).doStuff

I.e. you concatenate a source with a single tick and the ticksource. Now you have a guaranteed first tick as soon as there is demand, and then after a periodic source of ticks.

**Comment by francisdb** [comment]
That indeed sounds like a better plan, thanks.

**Comment by patriknw** [comment]
great idea, cookbook?

**Comment by drewhk** [comment]
Well, it is more like a workaround, since this might skew the length of the first period.

**Comment by rkuhn** [comment]
I was wondering about the implementation dropping ticks when at the time of arrival there was no demand; it might be better to record the availability of a tick and deliver it when demand arrives. This “buffer” should be limited to size 1, though, in order to keep things sane.

**Comment by drewhk** [comment]
I am not sure. You can always add a buffer, but removing an internal one (when you *want* the dropping behavior) is not possible.

**Comment by rkuhn** [comment]
an external buffer would not be able to fix this particular issue, though; ticks may always arrive late in any case, so delaying them by up to one delay period does not sound terrible to me
